python/cudf/cudf/_lib/pylibcudf/io/types.pyx

```diff
@@ -108,3 +147,85 @@ cdef class SourceInfo:
                                                      c_buffer.shape[0]))
 
         self.c_obj = source_info(c_host_buffers)
+
+
+# Adapts a python io.IOBase object as a libcudf IO data_sink. This lets you
+# write from cudf to any python file-like object (File/BytesIO/SocketIO etc)
+cdef cppclass iobase_data_sink(data_sink):
+    object buf
+
+    iobase_data_sink(object buf_):
+        this.buf = buf_
+
+    void host_write(const void * data, size_t size) with gil:
+        if isinstance(buf, io.StringIO):
+            buf.write(PyMemoryView_FromMemory(<char*>data, size, PyBUF_READ)
+                      .tobytes().decode())
+        else:
+            buf.write(PyMemoryView_FromMemory(<char*>data, size, PyBUF_READ))
+
+    void flush() with gil:
+        buf.flush()
+
+    size_t bytes_written() with gil:
+        return buf.tell()
+
+
+cdef class SinkInfo:
+    """A class containing details on a source to read from.
+
+    For details, see :cpp:class:`cudf::io::sink_info`.
+
+    Parameters
+    ----------
+    sinks : List[Union[str, os.PathLike,
+                        io.BytesIO, io.IOBase,
+                        io.StringIO, io.TextIOBase]]
+
+        A homogeneous list of sinks (this can be a string filename,
+        bytes, or one of the Python I/O classes) to read from.
+
+        Mixing different types of sinks will raise a `ValueError`.
+    """
+
+    def __init__(self, list sinks):
+        cdef vector[data_sink *] data_sinks
+
+        cdef vector[string] paths
+        if isinstance(sinks[0], io.StringIO):
+            data_sinks.reserve(len(sinks))
+            for s in sinks:
+                self.sink_storage.push_back(
+                    unique_ptr[data_sink](new iobase_data_sink(s))
+                )
+                data_sinks.push_back(self.sink_storage.back().get())
+            self.c_obj = sink_info(data_sinks)
+        elif isinstance(sinks[0], io.TextIOBase):
+            data_sinks.reserve(len(sinks))
+            for s in sinks:
+                # Files opened in text mode expect writes to be str rather than
+                # bytes, which requires conversion from utf-8. If the underlying
+                # buffer is utf-8, we can bypass this conversion by writing
+                # directly to it.
+                if codecs.lookup(s.encoding).name not in {"utf-8", "ascii"}:
+                    raise NotImplementedError(f"Unsupported encoding {s.encoding}")
+                self.sink_storage.push_back(
+                    unique_ptr[data_sink](new iobase_data_sink(s.buffer))
+                )
+                data_sinks.push_back(self.sink_storage.back().get())
+            self.c_obj = sink_info(data_sinks)
+        elif isinstance(sinks[0], io.IOBase):
```

Comment:
Is this branch supposed to be exactly the same as the `io.StringIO` case? If so would there be a way to combine this?

Reply:
Oh, I copied this verbatim, but this should be combinable.

I've updated it in the last commit.